src/reducers/console.js

```diff
@@ -35,20 +33,62 @@ export default function console(stateIn, {type, payload, meta}) {
         ),
       );
     case 'EVALUATE_CONSOLE_ENTRY':
-      return payload.trim(' ') === '' ? state : state.set(
-        meta.key,
-        new ConsoleEntry({expression: payload}),
+      return payload.trim(' ') === '' ? state : state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({expression: payload}),
       );
     case 'CLEAR_CONSOLE_ENTRIES':
       return initialState;
+    case 'CONSOLE_CHANGE':
+      return state.set('currentInputValue', payload.value);
     case 'CONSOLE_LOG_PRODUCED':
-      return state.set(
-        meta.key,
-        new ConsoleEntry({
+      return state.updateIn(
+        ['history', meta.key],
+        () => new ConsoleEntry({
           value: payload.value,
           evaluatedByCompiledProjectKey: payload.compiledProjectKey,
         }),
       );
+    case 'NAVIGATE_CONSOLE_HISTORY': {
+      const relevantHistory = state.history.toList().filter(
+        entry => entry.expression !== null,
+      );
```

Comment:
I’d be inclined to extract this into a selector using `reselect` to cache the last value so that we don’t have to recalculate it on every navigation unless the console history itself has changed.

Reply:
I think recalculation fine here - caching would complicate the code and I don't expect performance to be an issue. If performance measurements indicated that the ui was noticeable less responsive due to this (which I'm aware is extra important for students using old browsers and hardware) I'd happily look for such a solution, but I'll pass on implementing that for now.

Reply:
I don’t necessarily agree that it would complicate the code on balance—more indirection, yes, but also more self-describing code (via the selector’s name) and using a pattern that’s standard in the codebase. BUT I am receptive to the argument that it’s a premature optimization 🙂